File: c/src/servirtium.c
/* servirtium.c — the C client's implementation: the aether_vcr_embed_* ABI
 * declared once, plus the ergonomic layer servirtium.h promises.
 *
 * Carries NO record/replay logic. Every function here either forwards to the
 * engine, converts an integer, or handles string ownership. If you find
 * yourself wanting to parse a tape or match a request in this file, stop —
 * that is core/vcr.ae's job.
 */
#include "servirtium.h"

#include <stdlib.h>
#include <string.h>

/* ---- the raw C ABI (core/embed.ae) ------------------------------------- */
/* Declared here rather than in the public header: a consumer programs against
 * the sv_* surface, and these symbols are an implementation detail of the
 * seam. Handles are void* — NULL from an open means failure. Returned char*
 * are caller-owned and NUL-terminated; free with aether_vcr_embed_free_string. */

typedef void *sv_handle;

extern sv_handle aether_vcr_embed_open_playback(const char *label, const char *tape,
                                                const char *host, int port);
extern sv_handle aether_vcr_embed_open_playback_url(const char *label, const char *url,
                                                    const char *host, int port);
extern sv_handle aether_vcr_embed_open_record(const char *label, const char *tape,
                                              const char *upstream_base,
                                              const char *host, int port);
extern int   aether_vcr_embed_start(sv_handle h);
extern void  aether_vcr_embed_stop(sv_handle h);
extern char *aether_vcr_embed_stop_and_flush(sv_handle h, const char *tape);
extern char *aether_vcr_embed_stop_and_flush_fail_if_changed(sv_handle h, const char *tape);
extern char *aether_vcr_embed_stop_and_flush_or_check(sv_handle h, const char *tape);

extern int   aether_vcr_embed_port(sv_handle h);
extern char *aether_vcr_embed_base_url(sv_handle h, const char *host);
extern int   aether_vcr_embed_tape_length(sv_handle h);
extern void  aether_vcr_embed_reset_cursor(sv_handle h);

extern char *aether_vcr_embed_last_error(sv_handle h);
extern int   aether_vcr_embed_last_kind(sv_handle h);
extern int   aether_vcr_embed_last_index(sv_handle h);
extern void  aether_vcr_embed_clear_last_error(sv_handle h);

extern char *aether_vcr_embed_redact(sv_handle h, int field, const char *pat, const char *repl);
extern char *aether_vcr_embed_unredact(sv_handle h, int field, const char *pat, const char *repl);
extern char *aether_vcr_embed_normalize_whole_tape(sv_handle h, const char *pat, const char *name);
extern char *aether_vcr_embed_redact_whole_tape(sv_handle h, const char *pat, const char *repl);
extern char *aether_vcr_embed_remove_header(sv_handle h, int field, const char *name);
extern char *aether_vcr_embed_note(sv_handle h, const char *k, const char *v);
extern char *aether_vcr_embed_static_content(sv_handle h, const char *mount, const char *dir);
extern char *aether_vcr_embed_untaped(sv_handle h, const char *path);
extern char *aether_vcr_embed_match_header(sv_handle h, const char *name);
extern void  aether_vcr_embed_strict_ignore_common_headers(sv_handle h);
extern void  aether_vcr_embed_set_strict_headers(sv_handle h, int on);
extern void  aether_vcr_embed_set_match_json_body(sv_handle h, int on);
extern void  aether_vcr_embed_set_match_multiple(sv_handle h, int on);
extern void  aether_vcr_embed_indent_code_blocks(sv_handle h);
extern void  aether_vcr_embed_emphasize_http_verbs(sv_handle h);
extern void  aether_vcr_embed_clear_redactions(sv_handle h);
extern void  aether_vcr_embed_clear_unredactions(sv_handle h);
extern void  aether_vcr_embed_clear_normalizations(sv_handle h);
extern void  aether_vcr_embed_clear_header_removals(sv_handle h);
extern void  aether_vcr_embed_clear_match_headers(sv_handle h);
extern void  aether_vcr_embed_clear_static_content(sv_handle h);
extern void  aether_vcr_embed_clear_untaped(sv_handle h);
extern void  aether_vcr_embed_clear_format_options(sv_handle h);

extern char *aether_vcr_embed_har_import(const char *har_path, const char *tape_path);
extern char *aether_vcr_embed_har_export(const char *tape_path, const char *har_path);

extern void  aether_vcr_embed_free_string(char *s);

/* ---- the running server ------------------------------------------------ */

struct sv_vcr {
    sv_handle h;
    char *host;   /* owned; base_url needs it after open */
    char *tape;   /* owned; the flush default */
};

/* Why the last open failed. Static, per the header's contract. */
static char sv_open_err[512] = "";

static void set_open_error(const char *msg) {
    if (msg == NULL) msg = "";
    strncpy(sv_open_err, msg, sizeof(sv_open_err) - 1);
    sv_open_err[sizeof(sv_open_err) - 1] = '\0';
}

const char *sv_open_error(void) { return sv_open_err; }

/* Duplicate a NUL-terminated string with malloc. NOT strdup: strdup is POSIX,
 * not ISO C99, so under a strict -std=c99 it isn't declared (and on MSVC it is
 * spelled _strdup). This header advertises plain C99 with no dependencies
 * beyond the engine .so, so the four lines are cheaper than a feature macro. */
static char *dup_string(const char *s) {
    size_t n;
    char *out;
    if (s == NULL) return NULL;
    n = strlen(s) + 1;
    out = (char *)malloc(n);
    if (out != NULL) memcpy(out, s, n);
    return out;
}

/* Adopt an engine char* as an sv_str. NULL becomes an empty (but valid) one. */
static sv_str adopt(char *raw) {
    sv_str s;
    if (raw == NULL) {
        s.ptr = NULL;
        s.len = 0;
        return s;
    }
    s.ptr = raw;
    s.len = strlen(raw);
    return s;
}

/* An sv_str carrying nothing — what every void-ish config call yields. */
static sv_str empty_str(void) {
    sv_str s;
    s.ptr = NULL;
    s.len = 0;
    return s;
}

void sv_free(sv_str s) {
    if (s.ptr != NULL) aether_vcr_embed_free_string(s.ptr);
}

/* Copy the engine's last_error for the static open-error slot, then release
 * it — the caller of an open gets a const char*, not an owned string. */
static void capture_open_error(sv_handle h, const char *fallback) {
    char *raw = (h == NULL) ? NULL : aether_vcr_embed_last_error(h);
    if (raw != NULL && raw[0] != '\0') {
        set_open_error(raw);
    } else {
        set_open_error(fallback);
    }
    if (raw != NULL) aether_vcr_embed_free_string(raw);
}

/* Wrap an opened handle: start it, and on success own a copy of host + tape.
 * On any failure the handle is stopped and NULL returned, so a caller never
 * has to clean up after a failed open. */
static sv_vcr *wrap_started(sv_handle h, const char *host, const char *tape,
                            const char *fail_msg) {
    struct sv_vcr *vcr;

    if (h == NULL) {
        set_open_error(fail_msg);
        return NULL;
    }
    if (aether_vcr_embed_start(h) < 0) {
        capture_open_error(h, fail_msg);
        aether_vcr_embed_stop(h);
        return NULL;
    }
    vcr = (struct sv_vcr *)calloc(1, sizeof(*vcr));
    if (vcr == NULL) {
        set_open_error("out of memory");
        aether_vcr_embed_stop(h);
        return NULL;
    }
    vcr->h = h;
    vcr->host = dup_string(host);
    vcr->tape = dup_string(tape);
    set_open_error("");
    return vcr;
}

sv_vcr *sv_playback_on(const char *tape_path, const char *host, int port) {
    sv_handle h = aether_vcr_embed_open_playback("", tape_path, host, port);
    return wrap_started(h, host, tape_path, "playback open failed");
}

sv_vcr *sv_playback(const char *tape_path) {
    return sv_playback_on(tape_path, "127.0.0.1", 0);
}

sv_vcr *sv_playback_url(const char *tape_url) {
    sv_handle h = aether_vcr_embed_open_playback_url("", tape_url, "127.0.0.1", 0);
    /* No local tape path: a URL-sourced tape has nothing to flush back to. */
    return wrap_started(h, "127.0.0.1", NULL, "playback-url open failed");
}

sv_vcr *sv_record_on(const char *tape_path, const char *upstream_base,
                     const char *host, int port) {
    sv_handle h = aether_vcr_embed_open_record("", tape_path, upstream_base, host, port);
    return wrap_started(h, host, tape_path, "record open failed");
}

sv_vcr *sv_record(const char *tape_path, const char *upstream_base) {
    return sv_record_on(tape_path, upstream_base, "127.0.0.1", 0);
}

/* ---- introspection ----------------------------------------------------- */

sv_str sv_base_url(sv_vcr *vcr) {
    if (vcr == NULL) return empty_str();
    return adopt(aether_vcr_embed_base_url(vcr->h,
                                           vcr->host ? vcr->host : "127.0.0.1"));
}

int sv_port(sv_vcr *vcr) {
    return (vcr == NULL) ? -1 : aether_vcr_embed_port(vcr->h);
}

int sv_tape_length(sv_vcr *vcr) {
    return (vcr == NULL) ? -1 : aether_vcr_embed_tape_length(vcr->h);
}

void sv_reset_cursor(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_reset_cursor(vcr->h);
}

/* ---- diagnostics ------------------------------------------------------- */

sv_outcome sv_last_kind(sv_vcr *vcr) {
    if (vcr == NULL) return SV_OK;
    return (sv_outcome)aether_vcr_embed_last_kind(vcr->h);
}

sv_str sv_last_error(sv_vcr *vcr) {
    if (vcr == NULL) return empty_str();
    return adopt(aether_vcr_embed_last_error(vcr->h));
}

int sv_last_index(sv_vcr *vcr) {
    return (vcr == NULL) ? -1 : aether_vcr_embed_last_index(vcr->h);
}

void sv_clear_last_error(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_clear_last_error(vcr->h);
}

/* ---- configuration ----------------------------------------------------- */

sv_str sv_redact(sv_vcr *vcr, sv_field field, const char *pattern,
                 const char *replacement) {
    if (vcr == NULL) return empty_str();
    return adopt(aether_vcr_embed_redact(vcr->h, (int)field, pattern, replacement));
}

sv_str sv_unredact(sv_vcr *vcr, sv_field field, const char *pattern,
                   const char *replacement) {
    if (vcr == NULL) return empty_str();
    return adopt(aether_vcr_embed_unredact(vcr->h, (int)field, pattern, replacement));
}

sv_str sv_normalize_whole_tape(sv_vcr *vcr, const char *pattern, const char *name) {
    if (vcr == NULL) return empty_str();
    return adopt(aether_vcr_embed_normalize_whole_tape(vcr->h, pattern, name));
}

sv_str sv_redact_whole_tape(sv_vcr *vcr, const char *pattern, const char *replacement) {
    if (vcr == NULL) return empty_str();
    return adopt(aether_vcr_embed_redact_whole_tape(vcr->h, pattern, replacement));
}

sv_str sv_remove_header(sv_vcr *vcr, sv_field field, const char *name) {
    if (vcr == NULL) return empty_str();
    return adopt(aether_vcr_embed_remove_header(vcr->h, (int)field, name));
}

sv_str sv_note(sv_vcr *vcr, const char *title, const char *body) {
    if (vcr == NULL) return empty_str();
    return adopt(aether_vcr_embed_note(vcr->h, title, body));
}

sv_str sv_static_content(sv_vcr *vcr, const char *mount_path, const char *fs_dir) {
    if (vcr == NULL) return empty_str();
    return adopt(aether_vcr_embed_static_content(vcr->h, mount_path, fs_dir));
}

sv_str sv_untaped(sv_vcr *vcr, const char *path) {
    if (vcr == NULL) return empty_str();
    return adopt(aether_vcr_embed_untaped(vcr->h, path));
}

sv_str sv_match_header(sv_vcr *vcr, const char *name) {
    if (vcr == NULL) return empty_str();
    return adopt(aether_vcr_embed_match_header(vcr->h, name));
}

void sv_strict_ignore_common_headers(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_strict_ignore_common_headers(vcr->h);
}

void sv_set_strict_headers(sv_vcr *vcr, int on) {
    if (vcr != NULL) aether_vcr_embed_set_strict_headers(vcr->h, on);
}

void sv_set_match_json_body(sv_vcr *vcr, int on) {
    if (vcr != NULL) aether_vcr_embed_set_match_json_body(vcr->h, on);
}

void sv_set_match_multiple(sv_vcr *vcr, int on) {
    if (vcr != NULL) aether_vcr_embed_set_match_multiple(vcr->h, on);
}

void sv_indent_code_blocks(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_indent_code_blocks(vcr->h);
}

void sv_emphasize_http_verbs(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_emphasize_http_verbs(vcr->h);
}

void sv_clear_redactions(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_clear_redactions(vcr->h);
}

void sv_clear_unredactions(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_clear_unredactions(vcr->h);
}

void sv_clear_normalizations(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_clear_normalizations(vcr->h);
}

void sv_clear_header_removals(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_clear_header_removals(vcr->h);
}

void sv_clear_match_headers(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_clear_match_headers(vcr->h);
}

void sv_clear_static_content(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_clear_static_content(vcr->h);
}

void sv_clear_untaped(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_clear_untaped(vcr->h);
}

void sv_clear_format_options(sv_vcr *vcr) {
    if (vcr != NULL) aether_vcr_embed_clear_format_options(vcr->h);
}

/* ---- shutdown ---------------------------------------------------------- */

/* Release the wrapper itself. The engine handle is already stopped by the
 * caller; this frees only what the C layer owns. */
static void free_wrapper(struct sv_vcr *vcr) {
    free(vcr->host);
    free(vcr->tape);
    free(vcr);
}

void sv_close(sv_vcr *vcr) {
    if (vcr == NULL) return;
    aether_vcr_embed_stop(vcr->h);
    free_wrapper(vcr);
}

/* The three flush variants differ only in which ABI entry point they call, so
 * they share one body. `which`: 0 = plain, 1 = fail_if_changed, 2 = or_check. */
static sv_str close_flushing(sv_vcr *vcr, const char *tape_path, int which) {
    const char *tape;
    char *err;

    if (vcr == NULL) return empty_str();
    tape = (tape_path != NULL) ? tape_path : vcr->tape;
    if (tape == NULL) tape = "";

    if (which == 1) {
        err = aether_vcr_embed_stop_and_flush_fail_if_changed(vcr->h, tape);
    } else if (which == 2) {
        err = aether_vcr_embed_stop_and_flush_or_check(vcr->h, tape);
    } else {
        err = aether_vcr_embed_stop_and_flush(vcr->h, tape);
    }
    free_wrapper(vcr);
    return adopt(err);
}

sv_str sv_close_and_flush(sv_vcr *vcr, const char *tape_path) {
    return close_flushing(vcr, tape_path, 0);
}

sv_str sv_close_and_flush_fail_if_changed(sv_vcr *vcr, const char *tape_path) {
    return close_flushing(vcr, tape_path, 1);
}

sv_str sv_close_and_flush_or_check(sv_vcr *vcr, const char *tape_path) {
    return close_flushing(vcr, tape_path, 2);
}

/* ---- converters -------------------------------------------------------- */

sv_str sv_har_import(const char *har_path, const char *tape_path) {
    return adopt(aether_vcr_embed_har_import(har_path, tape_path));
}

sv_str sv_har_export(const char *tape_path, const char *har_path) {
    return adopt(aether_vcr_embed_har_export(tape_path, har_path));
}

File: swift/Sources/Servirtium/Servirtium.swift
// Servirtium.swift — the Swift binding over the shared Aether VCR engine.
//
// Servirtium records an HTTP conversation to a human-readable markdown tape
// once, then replays it forever — offline, deterministic, git-diffable. Point
// your system-under-test at `vcr.baseUrl` and drive it with plain HTTP.
//
//   let vcr = try Vcr.playback("tapes/single_get.md")
//   defer { vcr.close() }
//   let body = try String(contentsOf: URL(string: vcr.baseUrl + "/ok")!)
//   XCTAssertEqual(vcr.lastKind, .ok)
//
// Swift calls the engine's flat C ABI (aether_vcr_embed_*) DIRECTLY through the
// CServirtiumVcr module map — no glue .c, no second copy of the marshalling
// rules to drift from core/embed.ae.
//
// This file carries NO record/replay logic: markdown parse/emit, the HTTP
// server, request matching, redactions and drift detection all live in the
// in-repo pure-Aether core/vcr.ae engine. What Swift adds is a `withPlayback`
// scoped form that always closes, typed enums, `throws` instead of NULL/error
// strings, and automatic caller-owned-string handling.
import CServirtiumVcr
import Foundation

/// Field selector for redactions / unredactions / header removals. Values
/// mirror the FIELD_* constants in core/vcr.ae.
public enum Field: Int32 {
    case path = 1
    case responseBody = 2
    case requestHeaders = 3
    case requestBody = 4
    case responseHeaders = 5
}

/// Per-dispatch outcome. Anything but `.ok` is a mismatch. Values mirror the
/// VCR_KIND_* constants in core/vcr.ae.
public enum Outcome: Int32 {
    case ok = 0
    case pathOrMethodDiff = 1
    case headerMissing = 2
    case headerValueDiff = 3
    case headerUnexpected = 4
    case tapeExhausted = 5
    case bodyDiff = 6
    case recordError = 7
}

/// Thrown when a server can't be opened/started, or when a config call or a
/// record flush reports an error.
public struct VcrError: Error, CustomStringConvertible {
    public let message: String
    public init(_ message: String) { self.message = message }
    public var description: String { "VcrError: \(message)" }
}

/// Adopt an owned char* from the ABI as a Swift String and release it — the
/// caller-owned-string discipline, discharged in exactly one place.
private func take(_ ptr: UnsafeMutablePointer<CChar>?) -> String {
    guard let ptr = ptr else { return "" }
    let s = String(cString: ptr)
    aether_vcr_embed_free_string(ptr)
    return s
}

/// A config call's owned error string; empty means OK, anything else throws.
private func check(_ ptr: UnsafeMutablePointer<CChar>?, _ what: String) throws {
    let err = take(ptr)
    if !err.isEmpty { throw VcrError("\(what): \(err)") }
}

/// A running VCR server. One handle per instance, so N servers can run
/// concurrently in one process — one per port.
public final class Vcr {
    private let handle: UnsafeMutableRawPointer
    private let host: String
    private let tape: String?
    private var closed = false

    private init(handle: UnsafeMutableRawPointer, host: String, tape: String?) {
        self.handle = handle
        self.host = host
        self.tape = tape
    }

    // ---- lifecycle -------------------------------------------------------

    /// Replay a Servirtium markdown tape from disk.
    public static func playback(_ tapePath: String, host: String = "127.0.0.1",
                                port: Int32 = 0) throws -> Vcr {
        let h = aether_vcr_embed_open_playback("", tapePath, host, port)
        return Vcr(handle: try started(h, "playback open failed: \(tapePath)"),
                   host: host, tape: tapePath)
    }

    /// Replay a tape fetched from a URL rather than the filesystem.
    public static func playbackUrl(_ tapeUrl: String, host: String = "127.0.0.1",
                                   port: Int32 = 0) throws -> Vcr {
        let h = aether_vcr_embed_open_playback_url("", tapeUrl, host, port)
        return Vcr(handle: try started(h, "playback-url open failed: \(tapeUrl)"),
                   host: host, tape: nil)
    }

    /// Record: forward each request to `upstreamBase`, return the live response
    /// to the SUT, and capture the exchange. Call `flush()` to write the tape —
    /// a plain `close()` discards the recording.
    public static func record(_ tapePath: String, upstreamBase: String,
                              host: String = "127.0.0.1", port: Int32 = 0) throws -> Vcr {
        let h = aether_vcr_embed_open_record("", tapePath, upstreamBase, host, port)
        return Vcr(handle: try started(h, "record open failed: \(tapePath)"),
                   host: host, tape: tapePath)
    }

    /// Replay `tapePath` for the duration of `body`, closing on every path
    /// including a thrown error.
    public static func withPlayback<T>(_ tapePath: String, host: String = "127.0.0.1",
                                       port: Int32 = 0,
                                       _ body: (Vcr) throws -> T) throws -> T {
        let vcr = try playback(tapePath, host: host, port: port)
        defer { vcr.close() }
        return try body(vcr)
    }

    /// Record to `tapePath` for the duration of `body`, then flush. If `body`
    /// throws, the recording is DISCARDED — a failed test shouldn't overwrite a
    /// good tape.
    public static func withRecord<T>(_ tapePath: String, upstreamBase: String,
                                     host: String = "127.0.0.1", port: Int32 = 0,
                                     _ body: (Vcr) throws -> T) throws -> T {
        let vcr = try record(tapePath, upstreamBase: upstreamBase, host: host, port: port)
        do {
            let result = try body(vcr)
            try vcr.flush()
            return result
        } catch {
            vcr.close()
            throw error
        }
    }

    /// Start an opened handle, cleaning up and throwing rather than handing back
    /// a dead server. Shared by every open above.
    private static func started(_ handle: UnsafeMutableRawPointer?,
                                _ failure: String) throws -> UnsafeMutableRawPointer {
        guard let handle = handle else { throw VcrError(failure) }
        if aether_vcr_embed_start(handle) < 0 {
            let detail = take(aether_vcr_embed_last_error(handle))
            aether_vcr_embed_stop(handle)
            throw VcrError(detail.isEmpty ? failure : "\(failure): \(detail)")
        }
        return handle
    }

    // ---- introspection ---------------------------------------------------

    /// Base URL the SUT should target, e.g. "http://127.0.0.1:54213".
    public var baseUrl: String { take(aether_vcr_embed_base_url(handle, host)) }

    /// The OS-resolved port the server is listening on.
    public var port: Int32 { aether_vcr_embed_port(handle) }

    /// Tape entry count (playback), or interactions captured so far (record).
    public var tapeLength: Int32 { aether_vcr_embed_tape_length(handle) }

    /// Rewind the playback cursor to the top of the tape.
    public func resetCursor() { aether_vcr_embed_reset_cursor(handle) }

    // ---- diagnostics -----------------------------------------------------

    /// Outcome of the most-recent dispatch. `.ok` means a clean match.
    public var lastKind: Outcome {
        Outcome(rawValue: aether_vcr_embed_last_kind(handle)) ?? .ok
    }

    /// True when the most-recent dispatch matched cleanly.
    public var matchedCleanly: Bool { lastKind == .ok }

    /// Most-recent dispatch diagnostic; "" when none flagged.
    public var lastError: String { take(aether_vcr_embed_last_error(handle)) }

    /// Tape index of the most-recent matched interaction, or -1.
    public var lastIndex: Int32 { aether_vcr_embed_last_index(handle) }

    public func clearLastError() { aether_vcr_embed_clear_last_error(handle) }

    // ---- config (each throws if the engine rejects it) -------------------

    public func redact(_ field: Field, _ pattern: String, _ replacement: String) throws {
        try check(aether_vcr_embed_redact(handle, field.rawValue, pattern, replacement),
                  "redact")
    }

    public func unredact(_ field: Field, _ pattern: String, _ replacement: String) throws {
        try check(aether_vcr_embed_unredact(handle, field.rawValue, pattern, replacement),
                  "unredact")
    }

    /// Correlate every match into {{name-N}} tokens (things that recur and must
    /// stay consistent across the whole tape — UUIDs, CSRF tokens).
    public func normalizeWholeTape(_ pattern: String, _ name: String) throws {
        try check(aether_vcr_embed_normalize_whole_tape(handle, pattern, name),
                  "normalizeWholeTape")
    }

    /// Collapse every match to one constant (variable-cardinality noise — dates).
    public func redactWholeTape(_ pattern: String, _ replacement: String) throws {
        try check(aether_vcr_embed_redact_whole_tape(handle, pattern, replacement),
                  "redactWholeTape")
    }

    public func removeHeader(_ field: Field, _ name: String) throws {
        try check(aether_vcr_embed_remove_header(handle, field.rawValue, name),
                  "removeHeader")
    }

    /// Require this request header's live value to equal the recorded one. Also
    /// suppresses the automatic full-block header match unless strict headers
    /// are explicitly on — surgical, not all-or-nothing.
    public func matchHeader(_ name: String) throws {
        try check(aether_vcr_embed_match_header(handle, name), "matchHeader")
    }

    public func note(_ title: String, _ body: String) throws {
        try check(aether_vcr_embed_note(handle, title, body), "note")
    }

    public func staticContent(_ mountPath: String, _ fsDir: String) throws {
        try check(aether_vcr_embed_static_content(handle, mountPath, fsDir),
                  "staticContent")
    }

    public func untaped(_ path: String) throws {
        try check(aether_vcr_embed_untaped(handle, path), "untaped")
    }

    public func strictIgnoreCommonHeaders() {
        aether_vcr_embed_strict_ignore_common_headers(handle)
    }

    public func setStrictHeaders(_ on: Bool = true) {
        aether_vcr_embed_set_strict_headers(handle, on ? 1 : 0)
    }

    /// Opt-in: a request body that differs byte-for-byte gets a second chance
    /// at semantic JSON equality (key order + whitespace ignored, array order
    /// significant). A non-JSON body always falls back to the byte-exact verdict.
    public func setMatchJsonBody(_ on: Bool = true) {
        aether_vcr_embed_set_match_json_body(handle, on ? 1 : 0)
    }

    /// Opt-in: search ALL interactions for one that fits and replay it WITHOUT
    /// consuming it, so repeated and out-of-order requests both match.
    public func setMatchMultiple(_ on: Bool = true) {
        aether_vcr_embed_set_match_multiple(handle, on ? 1 : 0)
    }

    public func indentCodeBlocks() { aether_vcr_embed_indent_code_blocks(handle) }
    public func emphasizeHttpVerbs() { aether_vcr_embed_emphasize_http_verbs(handle) }
    public func clearRedactions() { aether_vcr_embed_clear_redactions(handle) }
    public func clearUnredactions() { aether_vcr_embed_clear_unredactions(handle) }
    public func clearNormalizations() { aether_vcr_embed_clear_normalizations(handle) }
    public func clearHeaderRemovals() { aether_vcr_embed_clear_header_removals(handle) }
    public func clearMatchHeaders() { aether_vcr_embed_clear_match_headers(handle) }
    public func clearStaticContent() { aether_vcr_embed_clear_static_content(handle) }
    public func clearUntaped() { aether_vcr_embed_clear_untaped(handle) }
    public func clearFormatOptions() { aether_vcr_embed_clear_format_options(handle) }

    // ---- shutdown --------------------------------------------------------

    /// Stop the server, discarding any recording. Idempotent.
    public func close() {
        if closed { return }
        closed = true
        aether_vcr_embed_stop(handle)
    }

    /// Stop and write the captured tape. `tapePath` defaults to the one the
    /// recorder was opened with. Idempotent.
    public func flush(_ tapePath: String? = nil) throws {
        try flush(tapePath) { aether_vcr_embed_stop_and_flush($0, $1) }
    }

    /// As `flush`, but fails if the fresh recording differs from what is
    /// already on disk (drift detection for a checked-in tape).
    public func flushFailIfChanged(_ tapePath: String? = nil) throws {
        try flush(tapePath) { aether_vcr_embed_stop_and_flush_fail_if_changed($0, $1) }
    }

    /// As `flush`, but only CHECKS against an existing tape instead of
    /// overwriting it when one is already present.
    public func flushOrCheck(_ tapePath: String? = nil) throws {
        try flush(tapePath) { aether_vcr_embed_stop_and_flush_or_check($0, $1) }
    }

    /// The three flush variants differ only in the ABI call, so they share this.
    private func flush(
        _ tapePath: String?,
        _ call: (UnsafeMutableRawPointer, String) -> UnsafeMutablePointer<CChar>?
    ) throws {
        if closed { return }
        closed = true
        try check(call(handle, tapePath ?? tape ?? ""), "flush")
    }
}

// ---- converters (no server handle) --------------------------------------

/// HAR 1.2 capture -> Servirtium markdown tape.
public func harImport(_ harPath: String, _ tapePath: String) throws {
    try check(aether_vcr_embed_har_import(harPath, tapePath), "harImport")
}

/// Servirtium markdown tape -> HAR 1.2 JSON.
public func harExport(_ tapePath: String, _ harPath: String) throws {
    try check(aether_vcr_embed_har_export(tapePath, harPath), "harExport")
}
